third/nmh/man/repl.man
.\"
.\" %nmhwarning%
.\" $Id: repl.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH REPL %manext1% MH.6.8 [%nmhversion%]
.SH NAME
repl \- reply to a message
.SH SYNOPSIS
.in +.5i
.ti -.5i
repl
\%[+folder] \%[msg]
.br
\%[\-group] \%[\-nogroup]
\%[\-annotate] \%[\-noannotate]
.br
\%[\-cc\ all/to/cc/me] \%[\-nocc\ all/to/cc/me]
.br
\%[\-query] \%[\-noquery]
\%[\-form\ formfile]
.br
\%[\-format] \%[\-noformat]
\%[\-filter\ filterfile]
.br
\%[\-inplace] \%[\-noinplace]
\%[\-mime] \%[\-nomime]
.br
\%[\-fcc\ +folder]
\%[\-width\ columns]
.br
\%[\-draftfolder\ +folder] \%[\-draftmessage\ msg]
.br
\%[\-nodraftfolder]
\%[\-editor\ editor] \%[\-noedit]
.br
\%[\-whatnowproc\ program] \%[\-nowhatnowproc]
.br
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIRepl\fR may be used to produce a reply to an existing message.

In its simplest form (with no arguments), \fIrepl\fR will set up a
message\-form skeleton in reply to the current message in the current
folder, and invoke the whatnow shell.

In order to construct the message draft of the reply, \fIrepl\fR uses
a reply template to guide its actions.  A reply template is simply a
\fImhl\fR format file (see \fImh\-format\fR\0(5) for details).

If the switch `\-nogroup' is given (it is on by default), then \fIrepl\fR
will use the standard forms file \*(lqreplcomps\*(rq.  This will construct
a draft message that is intended to be sent only to the author of the
message to which you are replying.  If a file named \*(lqreplcomps\*(rq
exists in the user's nmh directory, it will be used instead of this
default forms file.

The default reply template \*(lqreplcomps\*(rq will direct \fIrepl\fR
to construct the reply message draft as follows:

.nf
.in 1i
To: <Mail\-Reply\-To> or <Reply\-To> or <From>
Subject: Re: <Subject>
In\-Reply\-To: Your message of <Date>.
.ti +\w'In\-Reply\-To: 'u
<Message\-Id>
.in .5i
.fi

where field names enclosed in angle brackets (<\ >) indicate the
contents of the named field from the message to which the reply is
being made.

If the switch `\-group' is given, then \fIrepl\fR will use the the
standard forms file \*(lqreplgroupcomps\*(rq.  This will construct a
draft message that is intended as a group or followup reply.  If a file
named \*(lqreplgroupcomps\*(rq exists in the user's nmh directory, it
will be used instead of this default forms file.

The default group reply template \*(lqreplgroupcomps\*(rq will direct
\fIrepl\fR to construct the reply message draft as follows:

.nf
.in 1i
To: <Mail\-Followup\-To>
Subject: Re: <Subject>
In\-Reply\-To: Message from <From> of <Date>.
.ti +\w'In\-Reply\-To: 'u
<Message\-Id>
.in .5i
.fi

or if the field <Mail\-Followup\-To> is not available:

.nf
.in 1i
To: <Mail\-Reply\-To> or <Reply\-To> or <From>
cc: <To> and <cc> and <personal address>
Subject: Re: <Subject>
In\-Reply\-To: Message from <From> of <Date>.
.ti +\w'In\-Reply\-To: 'u
<Message\-Id>
.in .5i
.fi

In any case, you may specify an alternate forms file with the switch
`\-form\ formfile'.

You may selectively remove addresses from this default with the
`\-nocc\ type' switch.  This switch takes an argument (all/to/cc/me)
which specifies who gets removed from the default \*(lqcc:\*(rq list of
the reply.  You may give this switch multiple times (with different
arguments) if you wish to remove multiple types of addresses.

The `\-query' switch modifies the action of `\-nocc\ type' switch by
interactively asking you if each address that normally would be placed in
the \*(lqTo:\*(rq and \*(lqcc:\*(rq list should actually be sent a copy.
This is useful for special\-purpose replies.  Note that the position of
the `\-cc' and `\-nocc' switches, like all other switches which take a
positive and negative form, is important.

Lines beginning with the fields \*(lqTo:\*(rq, \*(lqcc:\*(rq, and
\*(rqBcc:\*(rq will be standardized and have duplicate addresses removed.
In addition, the `\-width\ columns' switch will guide \fIrepl\fR's
formatting of these fields.

If the draft already exists, \fIrepl\fR will ask you as to the disposition
of the draft.  A reply of \fBquit\fR will abort \fIrepl\fR, leaving the
draft intact; \fBreplace\fR will replace the existing draft with a blank
skeleton; and \fBlist\fR will display the draft.

See \fIcomp\fR\0(1) for a description of the `\-editor' and `\-noedit'
switches.  Note that while in the editor, the message being replied
to is available through a link named \*(lq@\*(rq (assuming the default
\fIwhatnowproc\fR\0).  In addition, the actual pathname of the message is
stored in the environment variable \fB$editalt\fR, and the pathname of
the folder containing the message is stored in the environment variable
\fB$mhfolder\fR.

Although \fIrepl\fR uses a forms file to direct it how to construct
the beginning of the draft, it uses a message filter file to direct
it as to how the message to which you are replying should be filtered
(re\-formatted) in the body of the draft.  The filter file for \fIrepl\fR
should be a standard form file for \fImhl\fR, as \fIrepl\fR will invoke
\fImhl\fR to format the message to which you are replying.

The switches `\-noformat', `\-format', and `\-filter\ filterfile' specify
which message filter file to use.

If the switch `\-noformat' is given (it is the default), then the message
to which you are replying is not included in the body of the draft.

If the switch `\-format' is given, then a default message filter file
is used.  This default message filter should be adequate for most users.
This default filter \*(lqmhl.reply\*(rq is:

.nf
.in +.5i
.ne 10
.eo
.so %etcdir%/mhl.reply
.ec
.in -.5i
.fi

which outputs each line of the body of the message prefaced with the
\*(lq>\*(rq character and a space.

If a file named \*(lqmhl.reply\*(rq exists in the user's nmh directory,
it will be used instead of this form.  You may specify an alternate
message filter file with the switch `\-filter\ filterfile'.

Other reply filters are commonly used, such as:

.nf
.in +.5i
:
body:nocomponent,compwidth=9,offset=9
.in -.5i
.fi

which says to output a blank line and then the body of the message
being replied\-to, indented by one tab\-stop.  Another popular format
is:

.nf
.in +.5i
.ie n \{
message-id:nocomponent,\|nonewline,\\
formatfield=\*(lqIn message %{text},\ \*(rq \}
.el message-id:nocomponent,\|nonewline,\|formatfield=\*(lqIn message %{text},\ \*(rq
from:nocomponent,\|formatfield=\*(lq%(friendly{text}) writes:\*(rq
body:component=\*(lq>\*(rq,\|overflowtext=\*(lq>\*(rq,\|overflowoffset=0
.in -.5i
.fi

This message filter file cites the Message-ID and author of the message
being replied\-to, and then outputs each line of the body prefaced with
the \*(lq>\*(rq character.

To use the MIME rules for encapsulation, specify the `\-mime' switch.
This directs \fIreply\fR to generate an \fImhbuild\fR composition file.
Note that nmh will not invoke \fImhbuild\fR automatically, unless you
add this line to your \&.mh\(ruprofile file:
.sp
.in +.5i
automimeproc: 1
.in -.5i
.sp
Otherwise, you must specifically give the command
.sp
.in +.5i
What now? mime
.in -.5i
.sp
prior to sending the draft.

If the `\-annotate' switch is given, the message being replied\-to will
be annotated with the lines

     Replied:\ date
     Replied:\ addrs

where the address list contains one line for each addressee.
The annotation will be done only if the message is sent directly from
\fIrepl\fR.  If the message is not sent immediately from \fIrepl\fR,
\*(lqcomp\ \-use\*(rq may be used to re\-edit and send the constructed
message, but the annotations won't take place.  Normally annotations are
done inplace in order to preserve any links to the message.  You may use
the `\-noinplace' switch to change this.

The `\-fcc\ +folder' switch can be used to automatically specify a folder
to receive Fcc:s.  More than one folder, each preceded by `\-fcc' can
be named.

In addition to the standard \fImh\-format\fR\0(5) escapes, \fIrepl\fR
also recognizes the following additional \fIcomponent\fR escape:
.sp 1
.nf
.ta \w'Escape  'u +\w'Returns  'u
\fIEscape\fR	\fIReturns\fR	\fIDescription\fR
\fIfcc\fR	string	Any folders specified with `\-fcc\ folder'
.re
.fi

To avoid reiteration, \fIrepl\fR strips any leading `Re: ' strings from
the \fIsubject\fR component.

The `\-draftfolder\ +folder' and `\-draftmessage\ msg' switches invoke
the \fInmh\fR draft folder facility.  This is an advanced (and highly
useful) feature.  Consult the \fImh-draft\fR(5) man page for more
information.

Upon exiting from the editor, \fIrepl\fR will invoke the \fIwhatnow\fR
program.  See \fIwhatnow\fR\0(1) for a discussion of available
options.  The invocation of this program can be inhibited by using the
`\-nowhatnowproc' switch.  (In truth of fact, it is the \fIwhatnow\fR
program which starts the initial edit.  Hence, `\-nowhatnowproc' will
prevent any edit from occurring.)

.Fi
^%etcdir%/replcomps~^The standard reply template
^or <mh\-dir>/replcomps~^Rather than the standard template
^%etcdir%/replgroupcomps~^The standard `reply -group' template
^or <mh\-dir>/replgroupcomps~^Rather than the standard template
^%etcdir%/mhl.reply~^The standard message filter
^or <mh\-dir>/mhl.reply~^Rather than the standard filter
^$HOME/\&.mh\(ruprofile~^The user profile
^<mh\-dir>/draft~^The draft file
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Alternate\-Mailboxes:~^To determine the user's mailboxes
.Ps
^Current\-Folder:~^To find the default current folder
.Ps
^Draft\-Folder:~^To find the default draft\-folder
.Ps
^Editor:~^To override the default editor
.Ps
^Msg\-Protect:~^To set mode when creating a new message (draft)
.Ps
^fileproc:~^Program to refile the message
.Ps
^mhlproc:~^Program to filter message being replied\-to
.Ps
^whatnowproc:~^Program to ask the \*(lqWhat now?\*(rq questions
.Sa
mhbuild(1), comp(1), forw(1), send(1), whatnow(1), mh\-format(5)
.De
`+folder' defaults to the current folder
.Ds
`msg' defaults to cur
.Ds
`\-nogroup'
.Ds
`\-cc\ all'
.Ds
`\-noannotate'
.Ds
`\-nodraftfolder'
.Ds
`\-noformat'
.Ds
`\-inplace'
.Ds
`\-nomime'
.Ds
`\-noquery'
.Ds
`\-width\ 72'
.Co
If a folder is given, it will become the current folder.  The message
replied\-to will become the current message.
.Bu
If any addresses occur in the reply template, addresses in the template
that do not contain hosts are defaulted incorrectly.  Instead of using
the localhost for the default, \fIrepl\fR uses the sender's host.
Moral of the story: if you're going to include addresses in a reply
template, include the host portion of the address.

The `\-width columns' switch is only used to do address-folding; other
headers are not line\-wrapped.

If \fIwhatnowproc\fR is \fIwhatnow\fR, then \fIrepl\fR uses a built\-in
\fIwhatnow\fR, it does not actually run the \fIwhatnow\fR program.
Hence, if you define your own \fIwhatnowproc\fR, don't call it
\fIwhatnow\fR since \fIrepl\fR won't run it.

If your current working directory is not writable, the link named
\*(lq@\*(rq is not available.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/repl.c

/*
 * repl.c -- reply to a message
 *
 * $Id: repl.c,v 1.2 1999-05-06 16:08:46 ghudson Exp $
 */

#include <h/mh.h>


static struct swit switches[] = {
#define GROUPSW                0
    { "group", 0 },
#define NGROUPSW               1
    { "nogroup", 0 },
#define	ANNOSW                 2
    { "annotate", 0 },
#define	NANNOSW                3
    { "noannotate", 0 },
#define	CCSW                   4
    { "cc all|to|cc|me", 0 },
#define	NCCSW                  5
    { "nocc type", 0 },
#define	DFOLDSW                6
    { "draftfolder +folder", 0 },
#define	DMSGSW                 7
    { "draftmessage msg", 0 },
#define	NDFLDSW                8
    { "nodraftfolder", 0 },
#define	EDITRSW                9
    { "editor editor", 0 },
#define	NEDITSW               10
    { "noedit", 0 },
#define	FCCSW                 11
    { "fcc folder", 0 },
#define	FILTSW                12
    { "filter filterfile", 0 },
#define	FORMSW                13
    { "form formfile", 0 },
#define	FRMTSW                14
    { "format", 5 },
#define	NFRMTSW               15
    { "noformat", 7 },
#define	INPLSW                16
    { "inplace", 0 },
#define	NINPLSW               17
    { "noinplace", 0 },
#define MIMESW                18
    { "mime", 0 },
#define NMIMESW               19
    { "nomime", 0 },
#define	QURYSW                20
    { "query", 0 },
#define	NQURYSW               21
    { "noquery", 0 },
#define	WHATSW                22
    { "whatnowproc program", 0 },
#define	NWHATSW               23
    { "nowhatnowproc", 0 },
#define	WIDTHSW               24
    { "width columns", 0 },
#define VERSIONSW             25
    { "version", 0 },
#define	HELPSW                26
    { "help", 4 },
#define	FILESW                27
    { "file file", -4 },		/* interface from msh */

#ifdef MHE
#define	BILDSW                28
    { "build", -5 },			/* interface from mhe */
#endif

    { NULL, 0 }
};

static struct swit ccswitches[] = {
#define	CTOSW	0
    { "to", 0 },
#define	CCCSW	1
    { "cc", 0 },
#define	CMESW	2
    { "me", 0 },
#define	CALSW	3
    { "all", 0 },
    { NULL, 0 }
};

static struct swit aqrnl[] = {
#define	NOSW         0
    { "quit", 0 },
#define	YESW         1
    { "replace", 0 },
#define	LISTDSW      2
    { "list", 0 },
#define	REFILSW      3
    { "refile +folder", 0 },
#define NEWSW        4
    { "new", 0 },
    { NULL, 0 }
};

static struct swit aqrl[] = {
    { "quit", 0 },
    { "replace", 0 },
    { "list", 0 },
    { "refile +folder", 0 },
    { NULL, 0 }
};

short ccto = 0;		/* global for replsbr */
short cccc = 0;
short ccme = 0;
short querysw = 0;

short outputlinelen = OUTPUTLINELEN;
short groupreply = 0;		/* Is this a group reply?        */

int mime = 0;			/* include original as MIME part */
char *form   = NULL;		/* form (components) file        */
char *filter = NULL;		/* message filter file           */
char *fcc    = NULL;		/* folders to add to Fcc: header */


/*
 * prototypes
 */
void docc (char *, int);


int
main (int argc, char **argv)
{
    int	i, isdf = 0;
    int anot = 0, inplace = 1;
    int nedit = 0, nwhat = 0;
    char *cp, *cwd, *dp, *maildir, *file = NULL;
    char *folder = NULL, *msg = NULL, *dfolder = NULL;
    char *dmsg = NULL, *ed = NULL, drft[BUFSIZ], buf[BUFSIZ];
    char **argp, **arguments;
    struct msgs *mp = NULL;
    struct stat st;
    FILE *in;

#ifdef MHE
    int buildsw = 0;
#endif /* MHE */

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s: [+folder] [msg] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case GROUPSW:
		    groupreply++;
		    ccto = cccc = ccme = 1;
		    continue;
		case NGROUPSW:
		    groupreply = 0;
		    continue;

		case ANNOSW: 
		    anot++;
		    continue;
		case NANNOSW: 
		    anot = 0;
		    continue;

		case CCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    docc (cp, 1);
		    groupreply = 1;
		    continue;
		case NCCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    docc (cp, 0);
		    continue;

		case EDITRSW: 
		    if (!(ed = *argp++) || *ed == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nedit = 0;
		    continue;
		case NEDITSW:
		    nedit++;
		    continue;
		    
		case WHATSW: 
		    if (!(whatnowproc = *argp++) || *whatnowproc == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    nwhat = 0;
		    continue;
#ifdef MHE
		case BILDSW: 
		    buildsw++;	/* fall... */
#endif /* MHE */
		case NWHATSW: 
		    nwhat++;
		    continue;

		case FCCSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    dp = NULL;
		    if (*cp == '@')
			cp = dp = path (cp + 1, TSUBCWF);
		    if (fcc)
			fcc = add (", ", fcc);
		    fcc = add (cp, fcc);
		    if (dp)
			free (dp);
		    continue;

		case FILESW: 
		    if (file)
			adios (NULL, "only one file at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    file = path (cp, TFILE);
		    continue;
		case FILTSW:
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    filter = getcpy (etcpath (cp));
		    mime = 0;
		    continue;
		case FORMSW: 
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case FRMTSW: 
		    filter = getcpy (etcpath (mhlreply));
		    mime = 0;
		    continue;
		case NFRMTSW: 
		    filter = NULL;
		    continue;

		case INPLSW: 
		    inplace++;
		    continue;
		case NINPLSW: 
		    inplace = 0;
		    continue;

		case MIMESW:
		    mime++;
		    filter = NULL;
		    continue;
		case NMIMESW:
		    mime = 0;
		    continue;

		case QURYSW: 
		    querysw++;
		    continue;
		case NQURYSW: 
		    querysw = 0;
		    continue;

		case WIDTHSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    if ((outputlinelen = atoi (cp)) < 10)
			adios (NULL, "impossible width %d", outputlinelen);
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			adios (NULL, "only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
				    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW:
		    if (dmsg)
			adios (NULL, "only one draft message at a time!");
		    if (!(dmsg = *argp++) || *dmsg == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    if (msg)
		adios (NULL, "only one message at a time!");
	    else
		msg = cp;
	}
    }

    cwd = getcpy (pwd ());

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (file && (msg || folder))
	adios (NULL, "can't mix files and folders/msgs");

try_it_again:

#ifdef MHE
    strncpy (drft, buildsw ? m_maildir ("reply")
			  : m_draft (dfolder, NULL, NOUSE, &isdf), sizeof(drft));

    /* Check if a draft exists */
    if (!buildsw && stat (drft, &st) != NOTOK) {
#else
    strncpy (drft, m_draft (dfolder, dmsg, NOUSE, &isdf), sizeof(drft));

    /* Check if a draft exists */
    if (stat (drft, &st) != NOTOK) {
#endif /* MHE */
	printf ("Draft \"%s\" exists (%ld bytes).", drft, (long) st.st_size);
	for (i = LISTDSW; i != YESW;) {
	    if (!(argp = getans ("\nDisposition? ", isdf ? aqrnl : aqrl)))
		done (1);
	    switch (i = smatch (*argp, isdf ? aqrnl : aqrl)) {
		case NOSW: 
		    done (0);
		case NEWSW: 
		    dmsg = NULL;
		    goto try_it_again;
		case YESW: 
		    break;
		case LISTDSW: 
		    showfile (++argp, drft);
		    break;
		case REFILSW: 
		    if (refile (++argp, drft) == 0)
			i = YESW;
		    break;
		default: 
		    advise (NULL, "say what?");
		    break;
	    }
	}
    }

    if (file) {
	/*
	 * We are replying to a file.
	 */
	anot = 0;	/* we don't want to annotate a file */
    } else {
	/*
	 * We are replying to a message.
	 */
	if (!msg)
	    msg = "cur";
	if (!folder)
	    folder = getfolder (1);
	maildir = m_maildir (folder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (folder)))
	    adios (NULL, "unable to read folder %s", folder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    adios (NULL, "no messages in %s", folder);

	/* parse the message range/sequence/name and set SELECTED */
	if (!m_convert (mp, msg))
	    done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	if (mp->numsel > 1)
	    adios (NULL, "only one message at a time!");

	context_replace (pfolder, folder);	/* update current folder   */
	seq_setcur (mp, mp->lowsel);	/* update current message  */
	seq_save (mp);			/* synchronize sequences   */
	context_save ();			/* save the context file   */
    }

    msg = file ? file : getcpy (m_name (mp->lowsel));

    if ((in = fopen (msg, "r")) == NULL)
	adios (msg, "unable to open");

    /* find form (components) file */
    if (!form) {
	if (groupreply)
	    form = etcpath (replgroupcomps);
	else
	    form = etcpath (replcomps);
    }

    replout (in, msg, drft, mp, outputlinelen, mime, form, filter, fcc);
    fclose (in);

    if (nwhat)
	done (0);
    what_now (ed, nedit, NOUSE, drft, msg, 0, mp,
	    anot ? "Replied" : NULL, inplace, cwd);
    done (1);
}

void
docc (char *cp, int ccflag)
{
    switch (smatch (cp, ccswitches)) {
	case AMBIGSW: 
	    ambigsw (cp, ccswitches);
	    done (1);
	case UNKWNSW: 
	    adios (NULL, "-%scc %s unknown", ccflag ? "" : "no", cp);

	case CTOSW: 
	    ccto = ccflag;
	    break;

	case CCCSW: 
	    cccc = ccflag;
	    break;

	case CMESW: 
	    ccme = ccflag;
	    break;

	case CALSW: 
	    ccto = cccc = ccme = ccflag;
	    break;
    }
}
